pkg/util/node/node_test.go

```diff
@@ -259,3 +231,201 @@ func TestGetHostname(t *testing.T) {
+		},
+		{
+			name:     "IPv4-only, external-first",
+			nodeName: "node1",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node1",
+				},
+				Spec: v1.NodeSpec{},
+				Status: v1.NodeStatus{Addresses: []v1.NodeAddress{
+					{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
+					{Type: v1.NodeExternalIP, Address: "4.3.2.2"},
+					{Type: v1.NodeInternalIP, Address: "1.2.3.4"},
+				}},
+			},
+			expect: netutils.ParseIPSloppy("1.2.3.4"),
+		},
+		{
+			name:     "IPv4-only, no internal",
+			nodeName: "node1",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node1",
+				},
+				Spec: v1.NodeSpec{},
+				Status: v1.NodeStatus{Addresses: []v1.NodeAddress{
+					{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
+					{Type: v1.NodeExternalIP, Address: "4.3.2.2"},
+				}},
+			},
+			expect: netutils.ParseIPSloppy("4.3.2.1"),
+		},
+		{
+			name:     "dual-stack node",
+			nodeName: "node1",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node1",
+				},
+				Spec: v1.NodeSpec{},
+				Status: v1.NodeStatus{Addresses: []v1.NodeAddress{
+					{Type: v1.NodeInternalIP, Address: "1.2.3.4"},
+					{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
+					{Type: v1.NodeExternalIP, Address: "4.3.2.2"},
+					{Type: v1.NodeInternalIP, Address: "a:b::c:d"},
+					{Type: v1.NodeExternalIP, Address: "d:c::b:a"},
+				}},
+			},
+			expect: netutils.ParseIPSloppy("1.2.3.4"),
+		},
+		{
+			name:     "dual-stack node, different order",
+			nodeName: "node1",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node1",
+				},
+				Spec: v1.NodeSpec{},
+				Status: v1.NodeStatus{Addresses: []v1.NodeAddress{
+					{Type: v1.NodeInternalIP, Address: "1.2.3.4"},
+					{Type: v1.NodeInternalIP, Address: "a:b::c:d"},
+					{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
+					{Type: v1.NodeExternalIP, Address: "4.3.2.2"},
+					{Type: v1.NodeExternalIP, Address: "d:c::b:a"},
+				}},
+			},
+			expect: netutils.ParseIPSloppy("1.2.3.4"),
+		},
+		{
+			name:     "dual-stack node, IPv6-first, no internal IPv4",
+			nodeName: "node1",
+			node: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node1",
+				},
+				Spec: v1.NodeSpec{},
+				Status: v1.NodeStatus{Addresses: []v1.NodeAddress{
+					{Type: v1.NodeInternalIP, Address: "a:b::c:d"},
+					{Type: v1.NodeExternalIP, Address: "d:c::b:a"},
+					{Type: v1.NodeExternalIP, Address: "4.3.2.1"},
+					{Type: v1.NodeExternalIP, Address: "4.3.2.2"},
+				}},
+			},
+			expect: netutils.ParseIPSloppy("a:b::c:d"),
+		},
+	}
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			kubeClientSet := kubeclientfake.NewSimpleClientset()
+			if test.node != nil {
+				if _, err := kubeClientSet.CoreV1().Nodes().Create(context.Background(), test.node, metav1.CreateOptions{}); err != nil {
+					t.Error("create node error")
+				}
+			}
+			result := GetNodeIP(kubeClientSet, test.nodeName)
```

Comment:
Tell me if I am misunderstanding:

Testing through `GetNodeIP()` seems to give coverage of the `wait.ExponentialBackoff()` logic but ultimately calls `GetNodeHostIP()`.  This test doesn't actually exercise any of the retry logic, so you're moving test cases around to score coverage "points" but it's bogus coverage.

I think I'd rather see all the test cases in the ` GetNodeHostIPs()` (plural) test, with a subset of those in a test for `GetNodeHostIP()` (singular).  This function (`GetNodeIP()`) seems to only be called from 2 places and I think it would be better to move this function to be internal to both.  "A little copy is better than a little dep".

Reply:
Thanks for the suggestion, I did notice that GetNodeIP() ends up calling GetNodeHostIP(). My GetNodeIP() test case and TestGetNodeHostIPs() both cover GetNodeHostIP(). I wonder if it's possible to cover wait.ExponentialBackoff() over GetNodeHostIP() directly via TestGetNodeIP(). So I removed TestGetNodeHostIPs. Not sure if this is possible?

Reply:
I think this moves in the wrong direction.  Tests should focus on the lowest level possible.

I'd set this up as a comprehensive test of `GetNodeHostIPs()` - since that seems to be where the real work is happening.

Testing `GetNodeIP()` (if done at all) should focus on the logic that is unique to that function, which is the backoff (which is mostly boilerplate and does not really belong in this utils lib in the first place).

Reply:
I'm sorry, with all your replies to me, I'm a little bit confused about what you mean here. Do I wish I could move the tests for GetNodeIP() to the called function directory, or should the tests be concentrated at the lowest level possible. If it is the second, I will remove the test cases that are repeated with TestGetNodeHostIPs() in TestGetNodeIP(). Then TestGetNodeIP() will only focus on testing GetNodeIP(). Not sure if I misunderstood what you meant if I changed it like this. Or should I move the test of GetNodeIP() to the corresponding test function of the called function, such as pkg/util/node/node_test.go Test_detectNodeIP(). Sorry to bother you.